### ISSUE TITLE
Support arbitrary subresources for custom resources

### ISSUE THREAD

**Issue by nikhita** [kubernetes/kubernetes#72637]
Currently, custom resources support only `status` and `scale` subresources. The only way to use custom subresources with CRs is to use an aggregated apiserver.

Given that using CRDs is comparatively easier than setting up an aggregated apiserver, there have been requests [[1](https://github.com/kubernetes/enhancements/issues/571#issuecomment-450404761)][[2](https://github.com/kubernetes/apiextensions-apiserver/issues/33)] for supporting arbitrary subresources with CRDs.

I don't think we intended to support this or have this on our roadmap in the near future. But we don't have a formal consensus on this yet. I created this issue to have a discussion and gather consensus.

/sig api-machinery
/area custom-resources
/kind feature
/cc @sttts @liggitt @deads2k @lavalamp @mbohlool

**Comment by yue9944882** [comment]
other than __plain__ arbitrary subresources, we might also provide extensibility over subresources via webhook: 

- resolving subresources as non-REST endpoint (e.g. long-running websocket calls, like `pod/log`,`pod/exec`)
- convert operations over subresources to main resource (e.g. `pod/binding`,`namespace/finalize`)

**Comment by nilebox** [comment]
Having support for arbitrary CRD subresources (implemented via webhooks) would definitely be useful, having to switch to custom API server is very expensive.

This may also be needed to ease the transition from API server to CRDs for Kubernetes core resources (many of Kubernetes core resources have subresources)? /cc @thockin

**Comment by krmayankk** [comment]
Isnt the main purpose of sub-resources to do fine grained RBAC on specific fields and allow subparts of a resource to be updated without updating the spec and hence generation ? It would be good to know what specific feature we are intending to implement in CRD and how subresource fulfills that ?

**Comment by nilebox** [comment]
@krmayankk the main purpose of subresources is support for arbitrary custom operation different from standard CRUD, e.g. `pods/exec` or `pods/log`. These subresources don't update any part of the `Pod` spec. In other words, the `/status` subresource is just one of the usage examples rather than the only use case.

**Comment by liggitt** [comment]
it's not clear the existing nonstandard subresources like exec or log are a pattern we want to encourage. I would put this feature on the list of things to consider after v1 of CRDs

**Comment by krmayankk** [comment]
@liggitt  do you consider status and scale also as non standard or only exec and log ?

**Comment by liggitt** [comment]
> @liggitt do you consider status and scale also as non standard or only exec and log ?

No, status and scale have well defined regular behavior. 

I was referring to exec, log, portforward, etc.

**Comment by nilebox** [comment]
@liggitt `exec` and `log` may be a bad example then, but if `scale` is fine (which is still specific to `Deployment`), then it would still require support for arbitrary per-resource subresources, unlike `status` which applies to all resources?

**Comment by liggitt** [comment]
> if scale is fine (which is still specific to Deployment), then it would still require support for arbitrary per-resource subresources

scale is generic and applies to many different resource types:

https://github.com/kubernetes/kubernetes/blob/2c8b571d573d434687316c606fb5e20e63008345/test/integration/scale/scale_test.go#L64

and is already supported by custom resources: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#scale-subresource

**Comment by nilebox** [comment]
Well it's still specific to resources managing `Pod`s (or some sort of replicas), but I see your point, thanks :)

**Comment by itaysk** [comment]
I think it would be nice to be able to report arbitrary information through the pod object, and I actually heard this ask from few others as well. For example, to report the status of the work that the pod does, or a certain aspect of the pod health like request latency. Getting these KPIs quickly by kubectl-ing the pod can be nice for operators, and for tooling to build on.
I was thinking on experimenting with adding such user defined status information to the status sub-resource but I guess it's pointless since the core controllers will probably override the status on their turn. This made me think about custom subresources so I ended up here.

**Comment by gwkunze** [comment]
I'm not sure this belongs in this topic or not but what about other specific, generalizable subresources? One I can think of is something similar to the approval subresource of certificate signing requests. Together with the appropriate RBAC rules you could give permission for entities to create resources that require some form approval by other entities (automated or human)?

**Comment by ncdc** [comment]
cc @pbarker

**Comment by pmorie** [comment]
I think it would be extremely useful to develop a standard for what a `/logs` subresource should do; it is a common want to people building abstractions on top of Kubernetes to want to be able to get logs for things via kubectl. Examples: Knative services, Tekton pipelines, etc.

**Comment by fabianvf** [comment]
I've seen a few operators using imperative CRDs (like Backup + Restore) that struck me as a possible case where custom subresources might be cleaner. 

As for `logs` I would definitely like to have that added to the group of standardized subresources, it would be incredibly useful especially when users have access to the Custom Resources but not the operator (ie, the operator is cluster-scoped). @pmorie +1 to opening a separate request for that.

**Comment by evankanderson** [comment]
Is `/logs` going through the apiserver is a pattern we want to encourage? If so, it would certainly be nice to be able to ask any resource for its logs, but I wonder if a pointer is better than e.g. merging logs for all Pods in a Deployment (for example)

**Comment by liggitt** [comment]
> Is `/logs` going through the apiserver is a pattern we want to encourage?

It is not.

> If so, it would certainly be nice to be able to ask any resource for its logs, but I wonder if a pointer is better than e.g. merging logs for all Pods in a Deployment (for example)

A redirect is better conceptually, but quickly gets pretty deep in the weeds... determining a TLS-verifiable connection to a network endpoint visible to all kube API consumers is non-trivial for things running inside the cluster to do currently.

**Comment by nikhilk** [comment]
In the absence of a generalized feature, is api server aggregation the answer? Is there a canonical example?

As mentioned by @fabianvf backup and restore are good examples of custom operations.

**Comment by mhrivnak** [comment]
"restart" is another action that would be useful for many CRDs. It's an imperative action that can involve a procedure of varying complexity. It can be hard to express declaratively in a spec, because the desired end result is for the thing to be "running", which it probably already is.

It's easy to think of other custom actions that would make sense for particular use cases, especially when storing data. "rebalance", "vacuum", "verify", "repair", etc.

**Comment by ncdc** [comment]
@mhrivnak these can example actions can also be achieved using "request" CRDs, such as `RestartRequest`, `VerifyRequest`, etc.

**Comment by mhrivnak** [comment]
> @mhrivnak these can example actions can also be achieved using "request" CRDs, such as `RestartRequest`, `VerifyRequest`, etc.

Yes indeed, I think nearly any action mentioned on this issue could be expressed that way. In cases where you want to carefully track both the status of individual actions and also the history of actions, that probably makes the most sense. But sometimes you just want the ability to restart a thing without the overhead of introducing a new resource.

**Comment by nikhilk** [comment]
I agree with @mhrivnak ... I think there are reasonable scenarios where verbs can be converted resource nouns, but there also those situations where that is just forced/unnatural, and trying to fit to a rigid model. Its always odd to see API programming styles try to fit everything into verbs (pure rpc) or everything into nouns (pure rest).

On that note, its also interesting to hear about the ownership semantics behind these XRequest CRDs.

- Should the controller assume ownership and clean out after the request has been processed?

- Alternatively, should the creator continue to own the resource, so they can watch it for completion, but then also be responsible for deleting it?

- Or should they just watch it and eventually something (what?) garbage collects those instances?

**Comment by ncdc** [comment]
> Should the controller assume ownership and clean out after the request has been processed?

What we do in Velero is have the controller delete the request CR immediately if the request was processed successfully (alternatively, we could set an ownerReference from the request CR to the other resource, and let GC handle the deletion). If not, we keep it around for a brief period of time. We may incorporate status information from the request CR when describing the resource it's associated with. Then we have the controller delete it after the expiration time.

> Alternatively, should the creator continue to own the resource, so they can watch it for completion, but then also be responsible for deleting it?

I imagine this would end up with a lot of "garbage" objects that the user forgets to delete.

**Comment by nikhilk** [comment]
@ncdc -- you sort of just made my point that resources are unnatural fit for this use-case.

**Comment by ncdc** [comment]
How so?

**Comment by ncdc** [comment]
BTW, I'm not at all against arbitrary subresources. I'm pointing out there are ways to achieve what you need today with the current API ecosystem and features.

If we can make it super easy for people to dynamically register webhook handlers for arbitrary subresources, that would be really cool.

I do have a small concern, though: if, as a project author, you expect/require subresource webhooks for your project to function correctly, you are dependent on the end user to install the webhooks into their cluster. If they forget, or have problems, your project might behave incorrectly. This is unfortunately a trade-off with CRDs. They're much easier to use for basic functionality, but with an aggregated API server, it's much harder (or impossible) to have partial functionality deployed.

**Comment by nikhilk** [comment]
IMO, it would be great if API aggregation was simpler, so you can author/deploy the combination of CRDs, a controller and API server in an easy/reliable manner.

I might be missing something but at the end of the day, a user must also deploy the CRD and controller to have the basic functionality. Ideally the additional API server bit could be both simplified, and rely on exactly same deployment model.

**Comment by liggitt** [comment]
> a user must also deploy the CRD and controller to have the basic functionality. Ideally the additional API server bit could be both simplified, and rely on exactly same deployment model.

the synchronous webhook/server model is not very similar to either the CRD deployment (simple manifest, kube-apiserver handles all the availability/TLS/serving) or the controller (asynchronous, can use existing pod/client-go/service account credentials to reach the API)

**Comment by andyxning** [comment]
Should we support finalize sub resource for CRD?

**Comment by n3wscott** [comment]
This is an issue dear to my heart. We in Knative have pushed the limits on what we can do without upstream/first class support for duck types. It is time to work upstream to implement custom subresources.

Here is a 1-pager game plan:
 
https://docs.google.com/document/d/1b1hRpJpSUE3xIG0v2IHFD0NBaxvRBTBnw3An7oZBV9I/edit?usp=sharing

**Comment by russgold** [comment]
A general assumption with Kubernetes resources is that they represent at most one thing that can be scaled. Our project, however, actually has a main resource with multiple included clusters - the reasons for this are historic and are tied to a legacy system more than two decades old, so changing it doesn't appear possible.  For that reason, we would like to be able to have multiple `scale` resources, distinguished by name. As a workaround, we are building a main resource and associated resources, each of which represents a cluster that is an aspect of the main one. It _works_, but is more complex than we could wish.

**Comment by cndoit18** [comment]
Any updates?
